apps/greencheck/models/checks.py

```diff
@@ -579,6 +585,49 @@ def from_sitecheck(cls, sitecheck):
             green=True,
         )
 
+    @classmethod
+    def claim_via_carbon_txt(cls, domain):
+        """
+        Accept a domain, and try verifying it with the carbon.txt validator.
+
+        We use the carbon.txt validator library to look up the domain, and
+        if there is the matching
+        domain we were expecting, we create the domain as "green domain" associated
+        with the provider that initially requested the domain hash.
+        """
+
+        # first of all, if there is no domain hash for the domain exit early - there is no
+        # point in trying to claim a domain hash for a domain that no provider has requested
+        # a domain hash for.
+        from ..domain_check import GreenDomainChecker
+
+        checker = GreenDomainChecker()
+
+        domain_hash = ac_models.DomainHash.objects.filter(domain=domain)
+        from ..exceptions import NoMatchingDomainHash
+
+        if not domain_hash:
+            raise NoMatchingDomainHash
+
+        # if we have a domain hash for the domain, choose the most recent one, so we can compare it to
+        # what we find on a domain
+        domain_hash = domain_hash.order_by("-created").first()
+
+        matched_hash = checker.verify_domain_hash(domain, domain_hash=domain_hash.hash)
+
+        # We don't want to create duplicate domains for the same provider.
+        # If a domain already exists for the provider, we should return it instead,
+        # of creating a new domain
+        if matched_hash:
+            provider = domain_hash.provider
+
+            if existing_domain := GreenDomain.objects.filter(
```

Comment:
As discussed, we'll need to filter by the provider who is asserting ownership here, to stop "looking over the shoulder" attacks!